src/yarp.c

```diff
@@ -2046,6 +2051,37 @@ parse_expression_prefix(yp_parser_t *parser) {
     }
     case YP_TOKEN_KEYWORD_FALSE:
       return yp_node_false_node_create(parser, &parser->previous);
+    case YP_TOKEN_KEYWORD_FOR: {
+      yp_token_t for_keyword = parser->previous;
+      yp_node_t *index = parse_left_hand_side(parser, BINDING_POWER_INDEX, "Expected index after for.");
+
+      expect(parser, YP_TOKEN_KEYWORD_IN, "Expected keyword in.");
+      yp_token_t in_keyword = parser->previous;
+
+      yp_node_t *collection = parse_expression(parser, BINDING_POWER_COMPOSITION, "Expected collection.");
+
+      yp_token_t do_keyword;
+      if (accept(parser, YP_TOKEN_KEYWORD_DO)) {
+        do_keyword = parser->previous;
+      } else {
+        not_provided(&do_keyword, parser->previous.start);
+      }
+
+      yp_node_t *scope = yp_node_scope_create(parser);
+      yp_node_t *parent_scope = parser->current_scope;
+      parser->current_scope = scope;
+
+      accept(parser, YP_TOKEN_SEMICOLON);
```

Comment:
@dirceu and I added handling for semicolons as well as newlines (and added a test case for it). Let me know if that looks alright.